### PR TITLE
Add tools + AWSDeploy tool

### PR DESCRIPTION
**Pull request by MichaelDeng03**
* Add tool metaclass that provides string representation
  * String representation includes object name + function information. Function information includes parameters, type hints, and docstring
* Add rough draft of DeployToAWS tool

### CODE REVIEW

**Comment by kgang** [comment]
what's the status of the PR? i.e. what's working or broken?

**Comment by MichaelDeng03** [comment]
I left some comments in tools.py too
Nothing thats broken afaik, but still missing some functionality, primarily:
Haven't hit response with eval yet
Haven't updated prompts/operators to use tools well
Missing a smarter pipeline for aws deploy w/ file handling

**Comment by MichaelDeng03** [comment]
My development has mostly been contained in a jupyter notebook. I've copied over my most important work, but its not really integrated into the rest of concrete

**Comment by kgang** [comment]
Anything in particular I can look into?

**Comment by kgang** [comment]
think about using a metaclass that handles all the superclasses instantiations and at the outermost layer, puts "tool" inside as a field

**Comment by kgang** [comment]
support normal function calls to tools' actions but pydantic specs for tool action inputs are a great idea

**Comment by kgang** [comment]
it's weird but fine to have an agent write down files before copying them into docker

File: concrete/__main__.py [hunk 1]
```diff
@@ -4,15 +4,19 @@
...
-    print("Use: `python -m concrete <prompt>`")
+if len(sys.argv) < 2:
+    print("Use: `python -m concrete <prompt> [--deploy]`")
     sys.exit(1)
-input = sys.argv[1]
+
+prompt = sys.argv[1]
+deploy_flag = False
+if len(sys.argv) > 2:
+    deploy_flag = sys.argv[2] == '--deploy'
```
Comment (kgang) [inline concrete/__main__.py:14]: use argparse here lol
https://docs.python.org/3/howto/argparse.html

File: concrete/operator_responses.py [hunk 2]
```diff
@@ -27,22 +27,32 @@ def __repr__(self):
...
+    tools: List[Tool] = Field(description="List of tools")
+
+
+class ProjectFile(Tools):
     file_name: str = Field(description="File path relative to project root")
     file_contents: str = Field(description="File contents")
 
 
-class ProjectDirectory(Response):
+class ProjectDirectory(Tools):
```
Comment (kgang) [inline concrete/operator_responses.py:44]: did you look into trying to remove `tool_name` and `tool_call` in the inner layers?

**Comment by kgang** [comment]
your pr on this looks good. I'm going to work on figuring out if we can take out the eval and if there's a better way to specify params for the tool function calls

**Comment by kgang** [comment]
hm ok, let's try to merge

**Comment by kgang** [comment]
I have just a couple comments

**Review by kgang — COMMENTED** [review]
Minor comments

File: concrete/__main__.py (hunk 1 above)
Comment (MichaelDeng03) [reply concrete/__main__.py:14]: .... noobie moment

File: concrete/operator_responses.py (hunk 2 above)
Comment (MichaelDeng03) [reply concrete/operator_responses.py:44]: No I wasn't able to get around to that. I'll look into that before resubmitting a pr.

Comment (kgang) [reply concrete/operator_responses.py:44]: No worries, I'm pretty sure I have a good idea how. Let's get this merged asap without it [](edited-with-Axolo)

**Review by kgang — APPROVED** [review]
good work!

**Pull request merged**